docs/DesignDocs/NotesOnAttestationAPI.md

```diff
@@ -0,0 +1,359 @@
+
+- The OE SDK is designed for attester and verifier software to be TEE agnostic,
+so that the same source code can be built for and run on different platforms
+- But owners of attesters and verifiers define their security policies,
+and the policies are not TEE agnostic. For example:
+  - The policies define which TEEs, platforms and configurations are trusted
+  to run the attesters and verifiers.
+    - They also define security properties of these entities, e.g. keypairs
+    and certs, service provider ID / SPID for SGX EPID evidence, etc.
+    These security properties are required for attestation and secure
+    communication.
+  - The policies define the list of evidence formats acceptable to the verifiers.
+    - Many acceptable evidence format are TEE specific. Examples include
+    SGX location attestation evidence formats, SGX ECDSA evidence formats,
+    and SGX EPID evidence formats.
+- The OE SDK provides means for the attesters and verifiers to be built and
+configured for specified TEEs and platforms, and to generate or verify evidences
+in the list of acceptable formats as defined in the security policies.
+  - The OE SDK attestation API enables the attesters and verifiers to register
+  the plugins that are built, configured, and working
+  properly, and enumerate the supported list of evidence formats covered by
+  these plugins.
+  For a specific attester or verifier instance, the supported list
+  of evidence formats is not required to perfectly match the acceptable list
+  defined in the security policies.
+
+# Notes on SGX Plugins Design
+
+## Evidence Format, Plugin, and Plugin Library
+
+In the OE SDK attestation API described in
+[Custom Attestation Data Formats for Open Enclave V3 Updates](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation_V3.md), every evidence format is uniquely identified by a UUID. Enclave and host
+applications can discover formats supported for evidence verification.
+Enclave applications can select a format for evidence generation. For evidence
+generation, API `oe_get_evidence()` takes in an evidence format UUID along with
+other input parameters. Upon successful operation, `oe_get_evidence()` returns
+an evidence blob with the evidence format UUID included as a field in its
+header structure.
+
+In the OE SDK attestation plugin design, a plugin is a subsystem that exposes
+a set of API entry points for evidence generation or verification, as defined
+in this [header file](https://github.com/openenclave/openenclave/blob/master/include/openenclave/attestation/plugin.h).
+A plugin can either be an attester plugin, or a verifier plugin.
+Each plugin is associated with a single evidence format UUID. On the other hand,
+in theory multiple plugins can exist in a configuration for the same format,
+though in practice this is unlikely since there is no clear benefit for the
+added complexity in implementation.
```

Comment:
Agree with Simon. Today Azure has multiple services to verify SGX quote - PCK Cache and MAA service. The verifier plugins for both will have the same format ID but it's unlikely an application will have both registered at the same time. The statements look good to me. Perhaps minor changes to address Simon's concern.

Reply:
@CodeMonkeyLeet very good points on the semantics of UUID. I'll update the document to capture them. The RE is an interesting idea. It is somewhat similar to the QE (Quoting Enclave) we are using for remote attestation, which works even if the attester and verifier enclave reside on the same platform.

@yentsanglee just curious what is MAA service? How does the PCK Cache verify quote? the name indicates it would simply retrieve and cache PCK certs. Do these services use the OE SDK plugins for quote verification or use some other implementations?

This document is growing to become much bigger than I originally expected :-).